### PR TITLE
Store/Get items in/from Blink instead of always calling from Cache.

### PR DESCRIPTION
**Pull request by Daynnnnn**
Sometimes Statamic will make duplicate calls to the `getItem` method in `Statamic\Stache\Stores\BasicStore`. While using the cache will alleviate some of the slow down caused by these duplicate calls, utilising blink will make these calls much faster.

Without Blink:
Screenshot 2021-07-12 at 16 46 20

With Blink:
Screenshot 2021-07-12 at 16 56 28

### CODE REVIEW

**Comment by Daynnnnn** [comment]
Happy for this to be reviewed now, let me know if you want anything changing.

**Comment by michaelr0** [comment]
Thanks @Daynnnnn I had also just come across this issue today and logged it as https://github.com/statamic/cms/issues/3991 - I've closed that in favor of moving the discussion into your PR

On a fresh install of Statamic and multiple refreshes to get the cache populated the results are as follows.

Before:
```
8 x hit stache::items::collections::pages (0%) | 0μs
3 x hit stache::items::entries::pages::home (0%) | 0μs
5 x hit stache::items::collection-trees::pages::default (0%) | 0μs
```

After:
```
1 x hit stache::items::collections::pages (0%) | 0μs
1 x hit stache::items::entries::pages::home (0%) | 0μs
1 x hit stache::items::collection-trees::pages::default (0%) | 0μs
```

**Comment by michaelr0** [comment]
Might be worth changing this

```php
    protected function cacheItem($item)
    {
        $key = $this->getItemKey($item);

        $cacheKey = $this->getItemCacheKey($key);

        Blink::forget($cacheKey);
        Cache::forever($cacheKey, $item);
    }
```

To be this

```php
    protected function cacheItem($item)
    {
        $key = $this->getItemKey($item);

        $cacheKey = $this->getItemCacheKey($key);

        Blink::put($cacheKey, $item);
        Cache::forever($cacheKey, $item);
    }
```

So that this, pushes the item to the Blink cache if it has to load it from the file?

```php
    public function getItem($key)
    {
        $this->handleFileChanges();

        if (! $path = $this->getPath($key)) {
            return;
        }

        if ($item = $this->getCachedItem($key)) {
            return $item;
        }

        $item = $this->makeItemFromFile($path, File::get($path));

        $this->cacheItem($item);

        return $item;
    }
```

File: src/Stache/Stores/BasicStore.php [hunk 1]
```diff
@@ -52,11 +55,13 @@ protected function cacheItem($item)
 
         $cacheKey = $this->getItemCacheKey($key);
 
+        Blink::forget($cacheKey);
```
Comment (jasonvarga) [inline src/Stache/Stores/BasicStore.php:58]: As per the suggestion in the comment, it seems like a good idea to just put it back into the Blink cache, rather than forgetting it.

`Blink::put($cacheKey, $item);`

**Review by jasonvarga — CHANGES_REQUESTED** [review]

File: src/Stache/Stores/BasicStore.php (hunk 1 above)
Comment (Daynnnnn) [reply src/Stache/Stores/BasicStore.php:58]: Hey, I agree this way would be better, and I had tried to do this initially - my issue was it was causing all the tests in `tests/Feature/Entries/LocalizeEntryTest.php` to fail when it was set to put. I spent a while debugging it, but couldn't work out what was going on. I'm guessing when `fresh()` is called on the entry it clears the cache, but I couldn't find where it was doing that clear to put the blink clear with it - and was also worried there'd be similar issues elsewhere. Not sure if you, or anyone has any ideas of what could be happening?

**Comment by michaelr0** [comment]
@Daynnnnn Looks like you figured out the tests? Great stuff!

File: tests/Stache/BasicStoreTest.php [hunk 2]
```diff
@@ -35,16 +35,17 @@ public function it_gets_an_item_by_key()
...
+        config(['cache.default' => 'file']);
         \Illuminate\Support\Facades\Cache::clear();
 
         file_put_contents($this->tempDir.'/foo.yaml', '');
 
         $this->assertNotNull($one = $this->store->getItem('foo'));
         $this->assertNotNull($two = $this->store->getItem('foo'));
-        $this->assertNotSame($one, $two);
+
+        $this->assertSame($one, $two);
```
Comment (jasonvarga) [inline tests/Stache/BasicStoreTest.php:48]: This test was intentional. We want different instances every time. Please revert this change.

See https://github.com/statamic/cms/commit/20d2c4cc50d316368e888d0882e17f2b6c7dc6c2

**Review by jasonvarga — CHANGES_REQUESTED** [review]

File: tests/Stache/BasicStoreTest.php (hunk 2 above)
Comment (Daynnnnn) [reply tests/Stache/BasicStoreTest.php:48]: Hi, now getItem returns the item from blink (which is in-memory). We would expect the result to be exactly the same. I think that's right?

Comment (michaelr0) [reply tests/Stache/BasicStoreTest.php:48]: Try

```php
    /** @test */
    public function items_are_different_instances_every_time()
    {
        config(['cache.default' => 'file']); // Doesn't work when they're arrays since the object is stored in memory.
        \Illuminate\Support\Facades\Cache::clear();

        file_put_contents($this->tempDir.'/foo.yaml', '');

        $this->assertNotNull($one = $this->store->getItem('foo'));
        $this->store->forgetItem('foo');

        $this->assertNotNull($two = $this->store->getItem('foo'));
        $this->assertNotSame($one, $two);
    }
```

Comment (michaelr0) [reply tests/Stache/BasicStoreTest.php:48]: Could have the both tests? if its worth it?

```php
    /** @test */
    public function items_are_different_instances_every_time()
    {
        config(['cache.default' => 'file']); // Doesn't work when they're arrays since the object is stored in memory.
        \Illuminate\Support\Facades\Cache::clear();

        file_put_contents($this->tempDir.'/foo.yaml', '');

        $this->assertNotNull($one = $this->store->getItem('foo'));
        $this->store->forgetItem('foo');

        $this->assertNotNull($two = $this->store->getItem('foo'));
        $this->assertNotSame($one, $two);
    }

    /** @test */
    public function items_are_the_same_instance_every_time()
    {
        config(['cache.default' => 'file']); // Doesn't work when they're arrays since the object is stored in memory.
        \Illuminate\Support\Facades\Cache::clear();

        file_put_contents($this->tempDir.'/foo.yaml', '');

        $this->assertNotNull($one = $this->store->getItem('foo'));
        $this->assertNotNull($two = $this->store->getItem('foo'));
        $this->assertSame($one, $two);
    }
```

Comment (Daynnnnn) [reply tests/Stache/BasicStoreTest.php:48]: Hmmm, I'm not sure we'd want to clear the cache between the getItem calls - as I don't think this would happen in the actual app?

Comment (michaelr0) [reply tests/Stache/BasicStoreTest.php:48]: I'd imagine that in the actual app, you'd want the same instance to be returned, but I could be wrong.

https://github.com/statamic/cms/commit/a364f472e66c49118d9f08c0707b71de63a14190 Seems to suggest that would also be the desired outcome.

While https://github.com/statamic/cms/commit/20d2c4cc50d316368e888d0882e17f2b6c7dc6c2 specifically tests that they arent.

@jasonvarga are you able to elaborate on this?
Would you foresee any issues with clearing the store inbetween to force them to be different instances?

Comment (jasonvarga) [reply tests/Stache/BasicStoreTest.php:48]: I understand that with your blink changes, that it now returns the same instance every time.

I'm saying - we _need_ it to be a different instance every time, because of the reasoning explained in 20d2c4c. Your implementation will also need to be changed.

Think about Eloquent. Every time you do `Post::find(1)`, it gets a fresh version from the DB.

Maybe you can keep blink, but do a `clone` so it's a different instance.
i.e. `return clone Blink::once($cacheKey, ...`

Comment (Daynnnnn) [reply tests/Stache/BasicStoreTest.php:48]: Ahhh okay, just clocked it, that make more sense now. Cloning seems like the best way to do it, though, I believe this might also require some changes to other tests.

Seen as though the `array` cache driver is used in the tests, getCachedItem() will return the same object, as it's in memory. But if we use blink with clone, then getCachedItem() will return a different instance (like we'd expect).

For this reason, I think some unit tests which compare items from the stache would need changing from assertSame() to assertEquals(). Would you agree?

Screenshot 2021-07-15 at 16 06 24

Comment (jasonvarga) [reply tests/Stache/BasicStoreTest.php:48]: Yeah I think thats okay there

**Comment by Daynnnnn** [comment]
Some words on the changes I've made, let me know if you think any of this looks wrong:
- GraphQL tests - Switch to `withSomeOfArgs` and remove entry expectation, as the entry will be a cloned instance.
- BuilderTest - Set `selectedQueryColumns` on `$searchedEntry`. Due to cloning, the field being set on `$retrievedEntry` no longer affects `$searchedEntry`. Even though this isn’t the expected behaviour, it’s the easiest way to keep the test working.
- MountingTest -  Save collection before also saving entries so URI is updated on entries in `findByUri` call.
- UpdateTaxonomyTest - I’ve added comments, but to summarise, we forget blinked taxonomies on entryOne as the taxonomy has changed, but the list hasn’t, so the cached blink response is stale. Reload taxonomies 2 and 3 as the taxonomy list has changed since they were originally created.
- UserGroupTest, AugmentedTestCase, FeatureTest, EntryRepositoryTest - swap from assertSame to assertEquals, as we’re comparing clones
- EntryTest - Save entries/collections after changes are made, and reload them before comparisons

**Comment by Daynnnnn** [comment]
Hey 👋 Don't suppose there's anything I can do to help push this along? This is contributing to some serious performance issues in a couple production sites atm, which is worrying the rest of the team when starting new projects on V3 :(

**Comment by jasonvarga** [comment]
This will probably be merged after 3.2 is out.

**Comment by duncanmcclean** [comment]
In the meantime, you may be able to pull in the fix using [composer patches](https://rias.be/blog/using-composer-patches)?

**Comment by isarphilipp** [comment]
Thanks @jasonvarga for putting me here after I filed a similar PR with a similar solution because we're hitting the same problems like @Daynnnnn . Is there anything we can contribute here to make it reality?

We're seeing -30% response times on a heavy site with this change, so that's a real benefit :)

**Comment by jasonvarga** [comment]
I'm thinking this may introduce more of the same bug being encountered in #2993.
If I'm understanding what's happening in there correctly - there's state being set that hangs around in the queue's process. When another process updates the cache, the first process still has old state.

**Comment by Daynnnnn** [comment]
I'd need to have a test, but I believe you're right. If I'm also understanding the issue correctly, I guess we could make sure the Stache blink is empty when we initialize the bits in `Statamic\Stache\ServiceProvider`?

**Comment by jasonvarga** [comment]
I think the service provider only runs once per process, in Horizon. I could be wrong about that. But yes if there was _something_ we could use to trigger a blink clear - I think that'd probably do it.

**Comment by Daynnnnn** [comment]
Hey @jasonvarga, @Zakini has suggested using the `Queue::before()` method [here](https://laravel.com/docs/8.x/queues#job-events) in one of the service providers (probably Stache) to clear the blink cache before each `JobProcessing` event, which should ensure a clean Blink before each job. Do you think this'll be alright?

**Comment by jasonvarga** [comment]
It could be! I didn't know about that method. It's likely very helpful.

**Comment by Daynnnnn** [comment]
I can confirm using `Queue::before` with `Blink::flush()` resolves this issue locally!

**Comment by jasonvarga** [comment]
How are you testing that?

**Comment by Daynnnnn** [comment]
Attached a video below. But in words:

- Created a standard statamic install using composer create-project --prefer-dist statamic/statamic statamic-test, and applied this PR to it.
- Set the queue driver to redis
- Created a job which echos the content field.
- Created a command to dispatch the job.
- Started the worker
- Ran the test job, changed the content via the CMS, and ran the test job again

With the new changes, the second job outputs the expected results. While before, the second job would output the old content.

(if the video is low quality when you view, you should be able to download it, it takes a while for drive/youtube to process it)
https://drive.google.com/file/d/1cf0mu5-F7MbaF2AQkuqwWu9poz4P_5bp/view

**Comment by Daynnnnn** [comment]
My last commit, [Clear taxonomy blink on save](https://github.com/statamic/cms/pull/3984/commits/2bb4887db31854e73d3037d23248d9c5e0a50145), is in place as there seems to be an issue with the cloning in BasicStore and the blinks on collections/taxonomies in entries/terms. Imo, those blink caches should now be removed - but this will have a performance hit on the eloquent driver, so I've left in for now. If you'd like me to go ahead and remove them, let me know.

**Comment by FrittenKeeZ** [comment]
This is very promising in terms of performance!
I'd love to this being merged soon :)

**Comment by michaelr0** [comment]
Just checking back in on this, it appears to still be pending?
Anything we/others can do to help this along?

**Comment by Daynnnnn** [comment]
Anything extra I can add to give us more confidence with this?

**Comment by o1y** [comment]
This PR is related to Issue #7424.

While testing this patch it turned out, that it's not working as intended. It seems that responses are much faster, but in some situations Stache returns wrong cached data which makes the returned data in CP or in frontend unreliable. I would really appreciate it, if someone here with more experience on the Stache Driver can check this PR again, as it can be improving overall performance up to 50% 🚀

Is it maybe a better approach avoiding the duplicate calls instead of changing the way how they are being cached?

**Comment by Daynnnnn** [comment]
@o1y You got any steps to reproduce this? We've been running this PR on ~10 live sites for over a year now, and we haven't hit any issues with it so far

**Comment by edalzell** [comment]
@Daynnnnn there is a merge conflict here, you wanna rebase on 3.4 and resolve?

**Comment by Daynnnnn** [comment]
Rebased and resolved conflicts, could still do with some steps to reproduce any issues though

**Comment by xuneXTW** [comment]
@Daynnnnn We recently ran into issues with multisite enabled. There where some weird behaviours when you filled content in one site and switched to an other site to fill in contents for this site on a new entry. After removing the patch it worked again. Maybe you could check if there are some issues with switching sites on new entries.

**Comment by Daynnnnn** [comment]
Had a look at this today, while we're cloning the actual item (e.g. an instance of Statamic\Entries\Entry) when it comes out of Blink, we're not cloning the objects which are inside that item (e.g. the data class variable on Entry, which is a collection object). This is why you can have 2 separate objects affect each other.

From my pov there's 2 solutions:
1. Implement the __clone method on all the classes which can be pulled out of the stache via the getCachedItem method, and take actions in there to make sure all objects in the item are cloned. 
2. Serialize the items before putting them into Blink, then unseralize as we fetch - which would match how the underlying cache works.

Solution 2 is probably better IMO - it does have some issues in tests though when you're mocking collection structures, and causes a few other failures which would need solving.

If one of those (or some other solution!) can be decided on, I'm can implement.

**Comment by jasonvarga** [comment]
Closing this PR as some [upcoming performance improvements](https://github.com/statamic/cms/pull/9588) may solve this another way. Thanks for understanding.
